twenty-first/src/shared_math/x_field_element.rs

```diff
@@ -242,44 +224,52 @@ impl Display for XFieldElement {
 }
 
 impl FromVecu8 for XFieldElement {
-    fn from_vecu8(&self, bytes: Vec<u8>) -> Self {
+    fn from_vecu8(bytes: Vec<u8>) -> Self {
         // TODO: See note in BFieldElement's From<Vec<u8>>.
         let bytesize = std::mem::size_of::<u64>();
         let (first_eight_bytes, rest) = bytes.as_slice().split_at(bytesize);
         let (second_eight_bytes, rest2) = rest.split_at(bytesize);
         let (third_eight_bytes, _rest3) = rest2.split_at(bytesize);
 
-        let coefficient0 = self.coefficients[0].from_vecu8(first_eight_bytes.to_vec());
-        let coefficient1 = self.coefficients[0].from_vecu8(second_eight_bytes.to_vec());
-        let coefficient2 = self.coefficients[0].from_vecu8(third_eight_bytes.to_vec());
+        let coefficient0 = BFieldElement::from_vecu8(first_eight_bytes.to_vec());
+        let coefficient1 = BFieldElement::from_vecu8(second_eight_bytes.to_vec());
+        let coefficient2 = BFieldElement::from_vecu8(third_eight_bytes.to_vec());
         XFieldElement::new([coefficient0, coefficient1, coefficient2])
     }
 }
 
-impl PrimeField for XFieldElement {}
-
-impl IdentityValues for XFieldElement {
-    #[inline]
-    fn is_zero(&self) -> bool {
-        self == &Self::ring_zero()
+impl Zero for XFieldElement {
+    fn zero() -> Self {
+        Self {
+            coefficients: [BFieldElement::zero(); 3],
+        }
     }
 
-    #[inline]
-    fn is_one(&self) -> bool {
-        self == &Self::ring_one()
+    fn is_zero(&self) -> bool {
+        self.coefficients.iter().all(|c| c.is_zero())
     }
+}
 
-    #[inline]
-    fn ring_zero(&self) -> Self {
-        Self::ring_zero()
+impl One for XFieldElement {
+    fn one() -> Self {
+        Self {
+            coefficients: [
+                BFieldElement::one(),
+                BFieldElement::zero(),
+                BFieldElement::zero(),
+            ],
+        }
     }
 
-    #[inline]
-    fn ring_one(&self) -> Self {
-        Self::ring_one()
+    fn is_one(&self) -> bool {
+        self.coefficients[0].is_one()
+            & self.coefficients[1].is_zero()
+            & self.coefficients[2].is_zero()
```

Comment:
```suggestion
        self.coefficients[0].is_one()
            && self.coefficients[1].is_zero()
            && self.coefficients[2].is_zero()
```

Reply:
Rust overloads the operator `BitAnd`, *i.e.*, `&`, to support the usual meaning when applied to bools.

I think the differences between `&` and `&&` in rust 
 - `&&` is only a logical operator whereas `&` also has other meanings – bitwise and, borrow, *etc.*
 - `&&` is lazy whereas `&` is not.

Field elements are typically used for cryptography where timing attacks, which can result from lazy evaluation, can be fatal. I would suggest using `&` by default. If the application allows, another method using `&&` can be written for faster average run time. But I'm not sure that can even be achieved because the lazy evaluation will throw the branch predictor into disarray.

Reply:
> I would suggest using `&` by default.

Do you mean
1) in low-level arithmetic like `add()`, `mul()`, `is_one()`, `mod_pow_u32()`,
1) in high-level arithmetic like `Polynomial<T>` and `MPolynomial<T>`,
1) in other timing-critical code than arithmetic, e.g. in hash functions, and/or
1) in general?

I believe we can divide all arithmetic optimizations we've made so far into "using algebraic properties, but possibly or deliberately constant-time", "deliberately using short-circuiting" (especially in the case of polynomial arithmetic), and "using memory efficiently". In short, if `&` vs. `&&` makes a difference here, we might want to (a) re-visit a large number of optimizations made so far, since some might have a larger and more obvious signature than `is_one()`, and (b) evaluate the [`CMOVcc` instructions for x86_64](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywZ7HAGTwNMAHLuAEaYxCDSAA6oCoR2DC5uHnrRsbYCvv5BLKHh0laYNvFCBEzEBInunlyWmNbpDCVlBJmBIWERlqXllck15s2t2bkRAJSWqCbEyOwcAPRzANQATACkAKzYAGySiwC0KxvYAMzLi6vLAEKLXKsaAIKRJsGLaAzmi8o7IIsmO%2BfHAAiiw0qio4KoAH0IVCNHCNJD4bdjpc7o9nq8BB9lFxlgAOH4mXF4gHA0Ew6EQxHw6kaZGoh5op4vKgMRZMdDoCDqX47UiLACehJ2o32q2O2F5u1WAHYGfdForFvQCIsFO5CcTSYtVOyFL8tRdrgK9Qb8eL5UrFhB1Sxzst1p9iaKmPq/pI0bLAZ6HszFqz2Zzoa6CNzhdJBeHRXtxZL3ec5WirSrFiwTKrbZr8drda6zSSjYLTUTzSik0q8FQ1e4AZKcdnZZarUrbWKgU7S03FV7y4rW3n3Z6Zd7Gb6MQGOehIQB3PwEMJUOq0MNS/lCqXR2NShNdxYLdn264vcXApj7a2Rc/BUa95WYVXcmqvLiik86gB0qAAbgvaKhZ0YkLqsEEBfLsBwCjeZYPMm96Bto2oaCYpzvtOxBMJEkR%2BMAQHPBAyBcMWpxQbuqhcKh6GYdhuEgRyCEDiKQ4jvcTLjmyabLjy7prlGbZxv8ja3imkTEBgJg2FmBbtrmbpagAVEWA7Ehat4QCJYk2Pajr1niLpujsTE%2BuiLLsSYtCQsQmDoOJmBcOsACcK7cZGG58dugkwUqwmidZEn5jmxbyYpsmdrelm%2BbZDlqT54kEDejLDkZfoBhxFlWTZkIsLIwCoGwxACk5fIue6m4Su5iaeYq3kaQQkkBUp2YKSaDV4iplWLOFGVZYIOV5QV6m%2BXFhmjvcCxLPcCi2pgiwEAgTCqrqbyxOY%2BqoFW/BTIspx7MEhCLF%2BYgmJgCg/Ew/LBPyyD8ugIBomNd37gcZ5vKUfj6qcd4hmqeDAAwlZ4KIgiLLtBAKKQD1LAc6CYoITBvVtZxZR8sS/f9gOqiDCjvhDEM6oseD6sEokANaMPjgioJixCWQo0QMPgRj7Ydx2mgoCD/myoR/tOaJLaq3gAPIAOrYAASpCACy9xCAA0uGiFglSMJtSxjLLMcGyXH4tB%2BJgEBiNOTACgoN7rMxAadTMdmOVxzpufGHkKl5cFMMe0nFgZ0FO1VcHINDb7crWW4ih7Hpe7BqrINq%2BHQwOxGhyrEeLP77Z%2B0HZWnCrt77mgLBPPO7IvFDFrA4UTBmNN0MsEbpeLMAlnzWEM1zWyrv8i36D0Is0TzoIeBiL89O/v%2BQlwRABB5xo/IE5CJhD8Qr7tq7n4/sQVDc9RwEQOgJGj6qE%2BRIRb4HxoFEYVhgFb4LIvi1LsuLApEAz3PWDEKHoy7yNVo57l%2BfTVHCk1ibA%2BgcZAJdZoE3xvqYAJgyiGHnFZRYDBUD70pi/YePN2opgPmcAOUdxROAIQjaMrws7tX3JOGamA85MzcCzQw0NLIECmGyWa00aZmQICXUIogK6LEwEwZACBFhrWbtNecNDq4mlCL8S8RBFj/ExuDchSwO5dx7owWwA9vwYL3taDhtALBQMhDohe2oD7kVMevf81FJzjzzssT%2B3sOrHU4WfKigE7HXzFpLaWMsH7WhnqY9%2B8UWKJRGhcDW6wtYMB1v4fWtBDbG1NsxPm/o2SW0wJlbKuUwgFVtviUq/FpQVWcfuIQmB2GYEXJZBgMxC46P9CQA8WAFDIGIHgSIDQRFVjYU02gG9GZ4DzvQNgsMGjYywXBVQtB6r6TDruFMqhhEB11LGYOkg9JSkTs7B8Z1%2BGL2BDMleGDbGcm5LMghRDiJkOcSmN2p53EXxwlvM86yM6OKeZvPC00GKbNuUnCAZBXiHJ1AgE5a9BkvLwteHZipiBfMvnhJCKE0Ln2%2BSBKOcdHHv2GvcDg4xaCcHWLwTwHAtCkFQJwUWZgMyTGmNNSJPBSC1XJQS8YxMQDrCnkSjgkhSWaF4FSjgvATpT1ZVocYcBYBIFzlhegZAKD4V/nQDoxAuBbGOFPGghiwgnQgMEQVpBdrMHypwZlucxkEAFrEoUbLSBYGrkYcQ9r8CWSKD%2BE69rMCqEKOmWYzK5x1CNTrImZQBQuCwEaggHSWDmoJXwAwwAFAADU8CYGnALSIjB40yEECIMQ7ApB5vkEoNQRrdA1AMEYEAphzD6DwMEE6kBxioC6fEL1ewnAiK6XsegP5aAnmOKKuohQGgOHpr0aopAfC62GB0SQ9lSCpDiAIKdKQYiroYEMdoeQl0FCKAIJoPRXBVD0Aehox6Whzt3REfd3QKinr6F0QYN6cgLvsuMBQ9KZh6BjZU%2BNhLiUCvtcKutBAo7kU1e%2BDQ1oaX1sWLgQgzSmX8hcHnVVb8mWjF4BK9lpBOXcv0JwflpAyUUuFaKkA4rBUf2IxwZYIGKOcFw7R8Yq9Yj2EkEAA%3D%3D%3D) that the efficient prime-field arithmetic uses ([`csel` on aarch64](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywZ7HAGTwNMAHLuAEaYxCAAHBqkAA6oCoR2DC5uHnpxCbYCvv5BLKHhUZaY1lkMQgRMxAQp7p5cxaVJFVUEOYEhYXrmrbVpDT3V7XkFEgCUlqgmxMjsHAD08wDUAEwApACs2ABskksAtKub2ADMK0trKwBCS1xrGgCCMSbBS2gM5kvKuyBLJrsXE4AESWGlUVAhVAA%2BpDoRp4RooQi7icrvcni83gJPsouCsIr8THiIoCQWDYTDIUiEdSNCi0Y90c9XlQGEsmOh0BB1H9dqQlgBPQm7MYHNYnbC8vZrADsDIeS0VS3oBCWCnchOJpKWqnZCj%2BWsuNwFeoN%2BPF8qVSwg6pYFxWGy%2BxNFTH1/0k6NlQM9j2ZS1Z7M5MNdBG5wukgvDov24sl7oucvRVpVSxYJlVts1%2BO1utdZpJRsFpqJ5tRSaVeCoavcgMluOzsstVqVtrFwKdpabiq95cVrbz7s9Mu9jN9mIDHPQUIA7n4CGEqCVaGGpfyhVLo7GpQmu0tFuz7TdXuKQUwDtaYufgmNe8rMKruQ03lxRSedQA6VAANwXtFQs6MKF1WCCBvj2Q4BRvMtHmTe9A20bUNBMM532nYgmBiGI/GAICXggZAuGLM4oN3VQuFQ9DMOw3CQI5BCBxFIcRweJlxzZNNlx5d01yjNs4wBRtbxTGJiAwEwbCzAt21zN0tQAKiLAdiQtW8IBEsSbHtR16wiF03V2JifQxFl2JMWgoWITB0HEzAuA2ABOFduMjDc%2BO3QSYKVYTROsiT8xzYt5MU2TO1vSzfNshy1J88SCBvRlhyMv0Aw4iyrJsqEWFkYBUDYYgBScvkXPdTcJXcxNPMVbyNIISSAqU7MFJNBqIhUyqlnCjKssEHK8oK9TfLiwzRweRZlgeBRbUwJYCAQJhVV1d4EnMfVUCrfhpiWM59mCQgli/MQTEwBRfiYflgn5ZB%2BXQEB0TGu790OM93kqPx9TOO8QzVPBgAYSs8FEQQll2ggFFIB7lkOdAsUEJg3q284ss%2BBJfv%2BwHVRBhR3whiGdSWPB9WCUSAGtGHxwRUCxYhLIUOIGHwIx9sO47TQUBB/zZUI/2ndEltVbwAHkAHVsAAJShABZB4hAAaXDRDwSpWE2pYxkVhOTYrj8Wg/EwCAxGnJgBQUG8NmYgNOtmOzHK4503PjDyFS8uCmGPaTiwM6CnaquDkGht9uVrLcRQ9j0vdg1VkG1fDoYHYjQ5ViOln99s/aDsqzhV299zQFhnnndlXihi1gcwUQzGm6GWCN0ulmASz5rCGa5rZV3%2BRb9B6CWOJ50EPAxD%2Benf3/IS4IgAg8%2BifGFChEwh%2BIV921dz8f2IKhueo4CIHQEjR9VCeYkIt8D40CiMKwwCt8FkXxal2WlgUiACdn%2BfQ7GXeRqtHPcvz6ao4U9YWwPqHGQCXWaBNp51xMFUQw84rJLAYKgfelM55YDXtzPeM087nADlHcUTh8EI2jG8LO7V9yThmpgPOTM3As0MNDSyBBphslmtNGmZkCAl1COXJQSxMBMGQAgJYa1m7TXnNQ6uJpQh/EvEQJYAJMbgzIcsDuXce6MFsAPb8w8ebtRTBAdhtALDTyhNohe2oD7kTMevf81FJzj2wR/b2HVjocLPlRQC9jr5i0ltLGWD9rTPzMW/eKLFEojUuBrDYWsGA638PrWghtjam2Ynzf0bJLaYEytlXKYQCq23xKVfi0oKrOP3EITAbDMCLksgwWYhdtH%2BhIAeLAChkDEDwDEMowiqysKabQDejM8B53oGwWGZRsZ6LgqoWg9V9Jh13CmVQQiA66ljMHSQekpSJ2dg%2BM6fDF4ghmSvHRdjOTclmfgwhxFSHOJTG7U87iL44S3medZGcVhjCeZvPC00GKbNuUnAxl1Dk6gQCc9BtjL54WvDsxUxBvnQpAkhFCaFz4/JAlHOOny37DQeBwCYtBOAbF4J4DgWhSCoE4KLMwGYpgzGmpEngpBarkoJRMYmIANjRCJRwSQpLNC8CpRwXgJ1oisq0BMOAsAkC5ywvQMgFB8I/zoF0YgXBtgnGiDQIxYQToQGCIK0gu1mD5U4My3OYyCAC1iUKNlpAsDVyMOIe1%2BBLI2DwD%2BE69rMCqDLumOYzK5wlCNTrImVQBQuCwEaggHSWDmoJXwAwwAFAADU8CYGnALGIjAE0yEECIMQ7ApD5vkEoNQRrdANAMEYEAphzD6DwMEE6kAJioC6Ukb1%2BwnDCK6fsegP5aAnhOAcfYlRiDABdlUQRux9hz2Jog6cDB%2B1%2BBMKofYv0TCipKGXMoDh6Z9HqKQHwusRhdAaBkRIAhD3pHiFehgwxOjhAGDuj1AgWjVBvS%2Bpo77x1tFPU%2B7of6v2WD/Y%2B/I56JgKHpbMPQsbKkJsJcSgV9rhX1oIFHcimr3waGtDShtSxcCEGaUy/kLg86quIPaE4L5eASvZaQTl3L9CcH5aQMlFLhWipAOKwV78WMcBWChzjnA6N8YmKvBI9hJBAA)).

Reply:
Comparison is the most obvious candidate for a function having a sensitive time signature. I would argue that we should avoid variable-time running times for low-level arithmetic like `add` and `mul` but I'm not sure that is a matter of urgent concern.

For high-level arithmetic including `Polynomial` and `MPolynomial` – specifically when the size is not fixed at compile-time – striving for a constant-time implementation is just silly because the running time has to depend on the size of the data.

Reply:
Let's look at the timing for the low-level stuff subsequently, then.

Consider the `&` → `&&` suggestion retracted.